sycl/doc/extensions/Assert/SYCL_ONEAPI_ASSERT.asciidoc

```diff
@@ -0,0 +1,144 @@
+= SYCL_EXT_ONEAPI_ASSERT
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+// This is necessary for asciidoc, but not for asciidoctor
+:cpp: C++
+
+== Notice
+IMPORTANT: This specification is a draft.
+
+Copyright (c) 2021-2021 Intel Corporation. All rights reserved.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.
+GitHub does not render image icons.
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 3.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to
+a feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+== Introduction
+This extension adds the ability for device code to call the C++ `assert()`
+macro. The behavior of `assert()` in device code is similar to its behavior in
+host code. If the asserted condition is false, a message is printed to `stderr`
+and then the program aborts with `std::abort()`.
+
+The format of the assert message is unspecified, but it will always include the
+text of the failing expression, the values of the standard macros `+__FILE__+`
+and `+__LINE__+`, and the value of the standard variable `+__func__+`. If the
+failing assert comes from an `nd_range` `parallel_for` it will also include the
+global ID and the local ID of the failing work item.
+
+It is unspecified whether a failing `assert()` returns to its caller before the
+kernel terminates. If a failing call returns, the device code may need to
+continue execution without deadlocking for the assertion message to be printed
+or for `std::abort()` to be called.
+
+The `assert()` macro is defined in system include headers, not in SYCL headers.
+On most of systems it is `<cassert>` and/or `<assert.h>` header files.
+The user can disable assertions in device code by defining the `NDEBUG`
+preprocessor macro prior to including either of `<sycl.hpp>` and
```

Comment:
Are you planning on adding asserts inside the SYCL header files?

Reply:
Fallback implementation will require changes in DPCPP/SYCL RT i.e. both in source and headers. Does it answer your question or did I get it wrong?

Reply:
I think I understand.  Is the idea that you don't want to define any internal functions that the assert macro relies on if NDEBUG is defined?

Reply:
These internal functions are already defined in devicelib. They include wrappers, which implement `__assert_fail` and `_wassert`. They call `__devicelib_assert_fail`. Fallback version of the latter is going to be changed to raise a flag in program-scope variable.

Reply:
In that case I don't understand why the NDEBUG macro must be set before the sycl.hpp header.  In what way will sycl.hpp use NDEBUG?

Reply:
SYCL headers will use `NDEBUG` macro to determine if fallback implementation of `__devicelib_assert_fail` will be working.

In other words, the state of `NDEBUG` macro is stored as property of device binary image. This allows program manager to get to know if online-linking against fallback devicelib is required.
On the other hand, SYCL headers don't have access to device binary image descriptor out-of-the-box. Hence, in order to decide if wrapper (for built-ins and buffer approach) or synchronization kernel + host-task (for program-scope variable approach) is needed, the headers contain `#ifndef NDEBUG` here and there.